Login: New Prologue Screens

rachelmcr opened the pull request:
Fixes: https://github.com/woocommerce/woocommerce-ios/issues/3824

## Description

Updates the current Prologue screen with new illustrations in a carousel to promote the following features:

- Stats
- Order Management
- Products
- Reviews

## Changes

The new Prologue screen and carousel has a few parts:

* `LoginPrologueViewController.swift` is still the main VC, with an updated background image and colors, and unused methods removed. `LoginPrologueViewController.xib` was also updated with these changes. (The RBG values for the color **Gray 0** are also updated to the correct values for this shade.)
* `LoginProloguePageViewController.swift` is the UIPageViewController, with a custom page control, embedded as a child view of `LoginPrologueViewController`. (This is modeled after [the UIPageViewController in WordPressAuthenticator](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/develop/WordPressAuthenticator/Signin/LoginProloguePageViewController.swift).)
* `LoginProloguePages.swift` contains the carousel's unique content (images and text) in `LoginProloguePageType` and shared styles/layout in `LoginProloguePageTypeViewController`.

### Screenshots

/|Stats|Order Management|Products|Reviews
-|-|-|-|-
Light|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 32 39](https://user-images.githubusercontent.com/8658164/112639998-d3f08480-8e38-11eb-888d-560dd9477c59.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 32 41](https://user-images.githubusercontent.com/8658164/112640009-d652de80-8e38-11eb-8280-641c3178ad47.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 32 43](https://user-images.githubusercontent.com/8658164/112640012-d6eb7500-8e38-11eb-8efb-45a7f6490565.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 32 46](https://user-images.githubusercontent.com/8658164/112640015-d6eb7500-8e38-11eb-8341-7e5f3125d621.png)
Dark|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 33 06](https://user-images.githubusercontent.com/8658164/112640022-d9e66580-8e38-11eb-863d-01eb6799c7d0.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 33 08](https://user-images.githubusercontent.com/8658164/112640023-da7efc00-8e38-11eb-83a3-216adc75bdae.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 33 10](https://user-images.githubusercontent.com/8658164/112640027-db179280-8e38-11eb-9bd2-b78ab6c919ce.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 33 12](https://user-images.githubusercontent.com/8658164/112640032-db179280-8e38-11eb-8675-e347e266763c.png)


iPhone SE|iPhone 11|iPad Pro (11")
-|-|-
![Simulator Screen Shot - iPhone SE (2nd generation) - 2021-03-26 at 13 32 18](https://user-images.githubusercontent.com/8658164/112640047-df43b000-8e38-11eb-9724-a8dc37c44e32.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-03-26 at 13 32 39](https://user-images.githubusercontent.com/8658164/112640054-e10d7380-8e38-11eb-9290-d1ad0b00ab6c.png)|![Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-03-26 at 13 31 30](https://user-images.githubusercontent.com/8658164/112640066-e4a0fa80-8e38-11eb-8329-ad382dae4dc0.png)



## Testing

1. Start logged out of the app.
2. Build and run the app, and confirm the initial (prologue) screen shows a carousel containing four pages with the new illustrations.
3. Confirm the layout displays all of the images/text, on both small and large device sizes.
4. Confirm the carousel and illustrations are reversed if the app is using a right-to-left language (e.g. Arabic, Hebrew).

## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Add new prologue carousel image assets

rachelmcr committed:
Update prologue screen background and colors

rachelmcr committed:
Remove unused UITextViewDelegate and action handler

rachelmcr committed:
Add Prologue screen carousel with promotional screens

rachelmcr committed:
Tweak Prologue screen layout

rachelmcr committed:
Update image tests

rachelmcr committed:
Update WordPressAuthenticator version for prologue background fix

rachelmcr committed:
Fix Gray0 RGB values

rachelmcr committed:
Update WordPressAuthenticator to beta

rachelmcr committed:
Fine-tune prologue screen layout

rachelmcr committed:
Merge branch 'develop' into issue/3824-prologue-carousel

rachelmcr committed:
Update release notes

Garance91540 reviewed (changes requested):
Hi Rachel! Thank you for working on this, they look so much better than what we currently have <3

Looking at your screenshots, I realized I hadn't included designs for tablets...Oupsie, I just added them to the Figma file:

<img width="382" alt="Screen Shot 2021-03-25 at 4 23 57 PM" src="https://user-images.githubusercontent.com/28450084/112498278-a77c3000-8d86-11eb-8769-948acc3f1ace.png">

rachelmcr committed:
Update constraints for iPad design

rachelmcr commented:
Thanks @Garance91540! I updated the layout to match the iPad design as closely as I could. I had trouble exactly matching the design without really altering the iPhone layout, so let me know if this is ok or if it needs more work:

<img src="https://user-images.githubusercontent.com/8658164/112640664-71e44f00-8e39-11eb-8d3f-fb29d8df10f3.png" width="300px">

(I also updated all the screenshots in the PR description above, so they reflect the latest changes.)

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+    private var pages: [UIViewController] = []
+
+    private var pageControl: UIPageControl!
```
Can we get rid of this implicitly unwrapped optional? Maybe turning this into a lazy variable?

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        dataSource = self
+        delegate = self
+
+        LoginProloguePageType.allCases.forEach({ type in
+            pages.append(LoginProloguePageTypeViewController(pageType: type))
+        })
```
What do you think about changing this to something like:
```swift
let pages: [UIViewController]  = {
    LoginProloguePageType.allCases.map { LoginProloguePageTypeViewController(pageType: $0) }
}()
```

That way the pages array can become immutable as it is intended.

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+            pages.append(LoginProloguePageTypeViewController(pageType: type))
+        })
+
+        setViewControllers([pages[0]], direction: .forward, animated: false)
```
Maybe adding something like
```swift
if let firstPage = pages.first {
    setViewControllers([firstPage], direction: .forward, animated: false)
}
```
Could avoid any crash in the future

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        let newPageControl = UIPageControl()
+        newPageControl.currentPageIndicatorTintColor = .gray(.shade5)
+        newPageControl.pageIndicatorTintColor = .wooCommercePurple(.shade50)
+        newPageControl.transform = CGAffineTransform(scaleX: 0.8, y: 0.8)
```
What do you think about extracting this magic number into a constant enum, to be easily changed and referenced?
Also, is this change due to the design?

rachelmcr replied:
> Also, is this change due to the design?

Yeah, I was trying to size the page control according to the design. Is that ok? If we don't usually do that, and want to just have the default page control size, we can remove this line.

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+        NSLayoutConstraint.activate([
+            newPageControl.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            newPageControl.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -10)
```
Same, about moving this into the enum 😇

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -0,0 +1,112 @@
+                            didFinishAnimating finished: Bool,
+                            previousViewControllers: [UIViewController],
+                            transitionCompleted completed: Bool) {
+        let toVC = previousViewControllers[0]
```
I think this is worth to be included in the guard statement below.
```swift
guard let toViewController = previousViewControllers.first, ... {
}
```

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePages.swift:
```diff
@@ -0,0 +1,130 @@
+    private let titleLabel = UILabel()
+    private let imageView = UIImageView()
+
+    private var pageType: LoginProloguePageType!
```
I don't think this needs to be an Implicitly unwrapped optional, right?

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePages.swift:
```diff
@@ -0,0 +1,130 @@
+
+        // Reduce centerYAnchor constraint priority to ensure the bottom margin has higher priority, so stack view is fully visible on shorter devices
+        let verticalCentering = stackView.centerYAnchor.constraint(equalTo: view.centerYAnchor, constant: 103)
+        verticalCentering.priority = UILayoutPriority(999)
```
nit: you can do `verticalCentering.priority = required - 1` in case the default required priority changes from 1000 in the future.

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePages.swift:
```diff
@@ -0,0 +1,130 @@
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            verticalCentering,
+            stackView.bottomAnchor.constraint(lessThanOrEqualTo: view.bottomAnchor, constant: -57),
```
Probably worth moving all these magic numbers no a constant enum, and provide a comment if necessary.

Ecarrion reviewed (commented):
Nice work @rachelmcr!
I left you some suggestions to gain a bit more consistency with the rest of the project. 
Also, I must admit that seeing so many magic numbers makes me a bit anxious coz I can't be sure if it will scale well a cross all screen sizes 😁

rachelmcr commented:
Thanks for the review!

> Also, I must admit that seeing so many magic numbers makes me a bit anxious coz I can't be sure if it will scale well a cross all screen sizes 😁

I feel the same! I think in some of our other VCs there are similar magic numbers but they are less obvious because they're set in a `.xib` file, instead of in code. That said, as I work through your feedback, I'll try to think if there are ways to remove some of these magic numbers.

rachelmcr committed:
Move constants into an enum

rachelmcr committed:
Remove implicitly unwrapped optionals

Garance91540 reviewed (approved):
👏

rachelmcr committed:
Cleanup based on PR feedback

rachelmcr committed:
Merge branch 'develop' into issue/3824-prologue-carousel

rachelmcr commented:
@Ecarrion Ready for another look! I moved the layout constants into enums, but I couldn't come up with any better ways to avoid those constants without deviating further from the designs. If there's another way to think about the constraints that you think would be more reliable, I'm definitely open to trying it; I might be unaware of a better way to construct it.

I've tested across a variety of devices (and it's a bit simpler because we don't support landscape on phones), and so far I think this approach works ok. There are some layout issues with the very largest dynamic type font sizes, but no functional issues.

Ecarrion commented on WooCommerce/Classes/Authentication/Prologue/LoginProloguePageViewController.swift:
```diff
@@ -4,9 +4,11 @@ import UIKit
+    }()
 
-    private var pageControl: UIPageControl!
+    private lazy var pageControl = UIPageControl()
```
Can't this be `private let pageControl = UIPageControl()` and just configure it below, and not re-assign it?

rachelmcr replied:
🤦  Yep, I don't know why I didn't do it that way the first time around! I'll make that change.

Ecarrion reviewed (approved):
Looks great! Thanks for taking the time on this! 🚀

rachelmcr committed:
Configure page control as a constant
